unitTests/magicbox/QueryProcessorTest.ts

```diff
@@ -0,0 +1,176 @@
+import { QueryProcessor, IQueryProcessResult, ProcessingStatus } from '../../src/magicbox/QueryProcessor';
+
+function wait<T = void>(ms?: number, result?: T): Promise<T> {
+  return new Promise(resolve => setTimeout(() => resolve(result), ms));
+}
+
+function flat<T>(bumpyArray: T[][]): T[] {
+  return bumpyArray.reduce((resultArray, array) => [...resultArray, ...array], []);
+}
+
+function expectArrayEqual<T>(actualArray: T[], expectedArray: T[]) {
+  if (!expect(actualArray).not.toBeNull()) {
+    return;
+  }
+  if (!expect(actualArray.length).toEqual(expectedArray.length, 'Lengths are not equal')) {
+    return;
+  }
+  if (actualArray.length === 0) {
+    return;
+  }
+  actualArray.forEach((actual, i) => expect(actual).toEqual(expectedArray[i], `Values at ${i} are not equal`));
+}
+
+export function QueryProcessorTest() {
+  describe('QueryProcessor', () => {
+    it('does not throw an error when overriding nothing', () => {
+      expect(() => new QueryProcessor().overrideIfProcessing()).not.toThrow();
+    });
+
+    describe('calling processQueries', () => {
+      describe('without values', () => {
+        let queryProcessor: QueryProcessor<string>;
+        let returnedPromise: Promise<IQueryProcessResult<string>>;
+        beforeEach(() => {
+          queryProcessor = new QueryProcessor();
+          returnedPromise = queryProcessor.processQueries([]);
+        });
+
+        it('resolves immediately', async done => {
+          const resolveProcess = jasmine.createSpy('resolveProcess');
+          returnedPromise.then(resolveProcess);
+          await wait();
+          expect(resolveProcess).toHaveBeenCalled();
+          done();
+        });
+
+        it('resolves with Finished status', async done => {
+          expect((await returnedPromise).status).toEqual(ProcessingStatus.Finished);
+          done();
+        });
+
+        it('resolves with an empty array', async done => {
+          expectArrayEqual((await returnedPromise).results, []);
+          done();
+        });
+      });
+
+      const timeBetweenQueries = 10;
+      const resolveTimes = [0, timeBetweenQueries, timeBetweenQueries * 2, timeBetweenQueries * 3];
```

Comment:
In setting up the tests, we will need different resolve times. However, one thing to keep in mind is we do not actually want to wait for those times, but use jasmine.clock to jump to specific points in time, and doing our assertions then. This helps keep our tests running fast,

Reply:
I didn't know about `jasmine.clock`, I'm switching to use it instead!